File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        List<List<String>> arr = new ArrayList<>();

        for(String str: strs){
            List<String> ordstr = map.getOrDefault(ordered(str), new ArrayList<>());
            ordstr.add(str);
            map.put(ordered(str), ordstr);
        }

        for(String str : map.keySet()){
            arr.add(map.get(str));
        }

        return arr;
    }

    private String ordered(String str) {
        char[] chars = str.toCharArray();
        Arrays.sort(chars);
        return Arrays.toString(chars);
    }
}

File: 49-group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)

        for s in strs:
            val = [0]*26
            for ch in s:
                val[(ord(ch) - ord('a'))] += 1
            groups[tuple(val)].append(s)

        return list(groups.values())